### ISSUE TITLE
Bindkey aliasing without warnings

### ISSUE THREAD

**Issue by somini** [neomutt/neomutt#451]
* Why do you not like the current state?

When you have several bindkeys with the same prefix, the one with the shortest prefix is always executed, making all others unreachable. It's even worse because the order of bindkey definition matters.

Take the minimal muttrc alternatives:

    bind index g group-reply
    bind index gg first-entry

When you click `g`, nothing happens because the bindkey that is defined is the second one. `gg` gets executed.

    bind index gg first-entry
    bind index g group-reply

When you click `g`, a group reply is initiated. The first entry binding was lost.

* What would you like to see to change?

I don't think mutt need a Vim-like approach using timeouts, that is too complicated. But at the very least a warning message should be produced when you bind a key combination that aliases other.

**Comment by guyzmo** [comment]
I agree and think it's a good idea, and thought about that reading [this thread on the mutt-devel list](http://marc.info/?t=148806613700004&r=1&w=2), regarding rebinding the `abort` function, which I think should be a bindable like the others (looks like the OP doesn't agree).

If you think you can do it, we can help you to do that. Otherwise, I can take some time as soon as I've got some.

**Comment by somini** [comment]
I skimmed the code and it seems the location is here:
https://github.com/neomutt/neomutt/blob/master/keymap.c#L251
This seems to be some kind of hash map, but I have never seen something like this implemented in C, and variable names like `tmp` don't help.

On that location the binding to be created is located on the `s` variable, but I can't seem to find the old variable they are overwriting, not even using the `km_expand_key` function.

**Comment by guyzmo** [comment]
the `Keymaps[menu]` is a way to select the key binding given the `menu` that is opened (which is the first argument of a binding, and can be `index`, `pager`… Each binding table is a sorted linked list (as hinted by the function's comment), so then it iterates over the list: `tmp = tmp->next;`.

Then the key sequence looks like it's stored as a character array (technically a `keycode_t` array) `tmp->keys`. So basically, to check for existence of a key, you need to iterate over `Keymaps[menu]`, for each element check its `->keys` property to be matched against the keycode string you generate using `parsekeys()`.

So below is a pseudo code to find a a keybinding containing one key:

```
len = parsekeys (s, buf, MAX_SEQ);
key = buf[0]; // we want only one keycode for this algorithm
for (keybinding = Keymaps[menu]; keybinding != NULL; keybinding = keybinding->next) {
  for (key_pos=0; key_pos < keybinding->len; ++key_pos) {
    if (keybinding->keys[key_pos] == key) {
      // there's a key binding containing key in the sequence!
      return true;
    }
  }
}
return false;
```

I left it untyped, because I only want to show you the iteration logic to find something in the code. It's untested and just an output of my understanding of reading the `km_bind` function.

**Comment by somini** [comment]
Thanks for that explanation.

This is my minimal test muttrc:
```
# Implicit default binds:
#  bind index g group-reply
#  bind pager g group-reply
bind index g group-reply
bind index gg first-entry # WARN
bind pager gg top # WARN
bind pager g group-reply # WARN
```
Since I'm checking for different lengths so that overwriting doesn't produce a message, there are 3 error messages printed with that muttrc.

Here's my hack printf version, just two commits:
https://github.com/somini/neomutt/tree/devel/issue-451
I was printing the characters of the bind one by one as per your pseudo-code, but the function `km_expand_key` does a similar thing, but produces printable strings, no raw "\n" inbetween.

**Comment by guyzmo** [comment]
can you make a pull request to ease reviewing of your patches?

**Comment by somini** [comment]
Sure, I'll get to it later today.

**Comment by somini** [comment]
Closed on #470
